app/models/king.rb

```diff
@@ -1,19 +1,56 @@
 # rubocop:disable Metrics/AbcSize
 # rubocop:disable Style/SymbolProc
 class King < Piece
+  @rook_id = nil # used by piece to get castled rook ID
+
+  # ensure King didn't move itself into check
+  def finalize_move!(to_row, to_col)
+    return false if get_attackers(to_row, to_col).present?
+    move_to!(to_row, to_col)
+  end
+
   def move_legal?(to_row, to_col)
+    @rook_id = nil
     row_diff = (row - to_row).abs
     col_diff = (col - to_col).abs
-    row_diff <= 1 && col_diff <= 1
-    # add castling moves to the 'whitelist', then verify the validity of the castle
-    #   further down in the logic
+    return true if row_diff <= 1 && col_diff <= 1
+
+    # King castling?
+    return false unless row_diff.zero? && col_diff == 2
+    move_legal_castle?(to_col)
   end
 
-  # override to enable castling
-  def finalize_move!(to_row, to_col)
-    return false if get_attackers(to_row, to_col).present?
-    move_to!(to_row, to_col)
-    # for castling something like: move_the_rook if castling?
+  def move_obstructed?(_to_row, _to_col)
+    false
+  end
+
+  def castling?(to_row, to_col)
+    return "castle. rook: #{@rook_id}" if row == to_row && (col - to_col).abs == 2
```

Comment:
You might be able to decrease the instance variable's scope by using [memoization](http://gavinmiller.io/2013/basics-of-ruby-memoization/).

```ruby
def castling?
  return "castle. rook: #{get_castling_rook(to_col).id}" if row == to_row && (col - to_col).abs == 2
  "normal"
end

def get_castling_rook(to_col)
  @_castling_rook ||= {}
  @_castling_rook[to_col] ||= begin
    rook_col = col < to_col ? 7 : 0
    rook = game.piece_at(row, rook_col)
    return false unless rook && rook.moves.empty?
    return rook if rook.king_rook_path_clear?(col)
    false
  end
end
```

The first line `@_castling_rook ||= {}` sets `@_castling_rook` to an empty hash if it is currently `nil`, so it can store cached pieces based on the key value of `to_col`.

The second either returns what's currently at `@_castling_rook[to_col]`, or it calculates it inside the block, saves, the value, then returns it.

In that way, you can call the method again later and get the same rook as before.  And you don't need to reference the instance variable in more than one method.